File: 2020/23.jl
function main(inp)
    inp = split(inp, "")
    inp = parse.(Int, inp)
    append!(inp, collect(maximum(inp)+1:10^6))
    nxt = [-1 for x in inp]
    max = maximum(inp)
    for (i, a) in enumerate(inp)
        nxt[a] = inp[begin + i % length(inp)]
    end
    cur = inp[1]
    for i in 1:10^7
        vals = Int[]
        orig = cur
        for i in 1:3
            cur = nxt[cur]
            push!(vals, cur)
        end
        cur = nxt[cur]
        nxt[orig] = cur
        dest = orig
        while dest == orig || dest in vals
            dest -= 1
            if dest < 1
                dest = max
            end
        end
        nxt[vals[3]] = nxt[dest]
        nxt[vals[2]] = vals[3]
        nxt[vals[1]] = vals[2]
        nxt[dest] = vals[1]
    end
    println(nxt[1] * nxt[nxt[1]])
end

main("156794823")

File: 2020/23.py
def main(inp):
    inp = [int(x) - 1 for x in inp]
    inp.extend([*range(max(inp) + 1, 10 ** 6)])
    nxt = [None for x in inp]
    mx = max(inp) + 1
    for a, b in zip(inp, inp[1:] + inp[:1]):
        nxt[a] = b
    cur = inp[0]
    for _ in range(10 ** 7):
        vals = []
        orig = cur
        cur = nxt[cur]
        for _ in range(3):
            vals.append(cur)
            cur = nxt[cur]
        nxt[orig] = cur
        dest = orig
        while dest in {*vals, orig}:
            dest = (dest - 1 + mx) % mx
        vals = [dest, *vals, nxt[dest]]
        for a, b in zip(vals, vals[1:]):
            nxt[a] = b
    print((nxt[0] + 1) * (nxt[nxt[0]] + 1))

from aocd import get_data
main(get_data(day=23))
